core-primitives/attestation-handler/src/attestation_handler.rs

```diff
@@ -0,0 +1,648 @@
+};
+use sp_core::{blake2_256, Pair};
+use std::{
+	borrow::ToOwned,
+	format,
+	io::{Read, Write},
+	net::TcpStream,
+	println, str,
+	string::{String, ToString},
+	sync::Arc,
+	vec::Vec,
+};
+use substrate_api_client::{compose_extrinsic_offline, ExtrinsicParams};
+
+pub const DEV_HOSTNAME: &str = "api.trustedservices.intel.com";
+
+#[cfg(feature = "production")]
+pub const SIGRL_SUFFIX: &str = "/sgx/attestation/v4/sigrl/";
+#[cfg(feature = "production")]
+pub const REPORT_SUFFIX: &str = "/sgx/attestation/v4/report";
+
+#[cfg(not(feature = "production"))]
+pub const SIGRL_SUFFIX: &str = "/sgx/dev/attestation/v4/sigrl/";
+#[cfg(not(feature = "production"))]
+pub const REPORT_SUFFIX: &str = "/sgx/dev/attestation/v4/report";
+
+pub trait AttestationHandler {
+	fn perform_ra(
+		&self,
+		genesis_hash_slice: &[u8],
+		nonce: u32,
+		url_slice: &[u8],
+	) -> EnclaveResult<Vec<u8>>;
+
+	fn get_mrenclave(&self) -> EnclaveResult<[u8; MR_ENCLAVE_SIZE]>;
+
+	fn dump_ra_to_disk(&self) -> EnclaveResult<()>;
+}
+
+pub struct IasAttestationHandler<OCallApi, NodeMetadataRepository> {
+	ocall_api: Arc<OCallApi>,
+	node_metadata_repo: Arc<NodeMetadataRepository>,
+}
+
+impl<OCallApi, NodeMetadataRepository> AttestationHandler
+	for IasAttestationHandler<OCallApi, NodeMetadataRepository>
+where
+	OCallApi: EnclaveAttestationOCallApi,
+	NodeMetadataRepository: AccessNodeMetadata<MetadataType = NodeMetadata>,
+{
+	fn perform_ra(
+		&self,
+		genesis_hash_slice: &[u8],
+		nonce: u32,
+		url_slice: &[u8],
+	) -> EnclaveResult<Vec<u8>> {
+		// our certificate is unlinkable
+		let sign_type = sgx_quote_sign_type_t::SGX_UNLINKABLE_SIGNATURE;
+
+		let (_key_der, cert_der) = match self.create_ra_report_and_signature(sign_type, false) {
+			Ok(r) => r,
+			Err(e) => return Err(e),
+		};
+
+		info!("    [Enclave] Compose extrinsic");
+		let signer = match Ed25519Seal::unseal_from_static_file() {
+			Ok(pair) => pair,
+			Err(e) => return Err(EnclaveError::Crypto(e)),
+		};
+		info!("[Enclave] Restored ECC pubkey: {:?}", signer.public());
+
+		debug!("decoded nonce: {}", nonce);
+		let genesis_hash = hash_from_slice(genesis_hash_slice);
+		debug!("decoded genesis_hash: {:?}", genesis_hash_slice);
+		debug!("worker url: {}", str::from_utf8(url_slice).unwrap());
+
+		let (register_enclave_call, runtime_spec_version, runtime_transaction_version) =
+			match self.node_metadata_repo.get_from_metadata(|m| {
+				(
+					m.register_enclave_call_indexes(),
+					m.get_runtime_version(),
+					m.get_runtime_transaction_version(),
+				)
+			}) {
+				Ok(r) => r,
+				Err(e) => {
+					error!("Failed to get node metadata: {:?}", e);
+					return Err(EnclaveError::Sgx(sgx_status_t::SGX_ERROR_UNEXPECTED))
+				},
+			};
+
+		let call = match register_enclave_call {
+			Ok(c) => c,
+			Err(e) => {
+				error!("Failed to get the indexes for the register_enclave call from the metadata: {:?}", e);
+				return Err(EnclaveError::Sgx(sgx_status_t::SGX_ERROR_UNEXPECTED))
+			},
+		};
```

Comment:
Here we have to keep in mind that @echevrier is just about to merge PR https://github.com/integritee-network/worker/pull/1033 which fixed issues in this part of the code (or rather the original `attestation.rs`). We need to make sure that we get those fixes into this refactoring as well and that they don't get lost in the rebase/merge

Reply:
Thanks for the remark. I incorporated the changes in 8a47b5bcf503d355c6a3fa413e43c1974230653a 
Can't hurt if you double check.

Reply:
Yes, that's looking to to me 👍